### DISCUSSION TITLE
understanding the flashHeader

### DISCUSSION THREAD

**Discussion by erikjonkers** [rhulme/pico-flashloader#8] [General]
Hi,

The way I receive my new firmware is a bit different than the example application, so I need to create and write the header after I saved all the data in flash.
To fully understand the data in the header I read it back in app.c after flashing it. See here:

```
FB5ECD8EA952AEC5C83B0000D3F26AFA
00B5324B212058609868022188439860
D860186158612E4B0021996002215961
0121F02299502B491960012199603520
00F044F80222904214D00621196600F0
34F8196E01211966002018661A6600F0
2CF8196E196E196E052000F02FF80121
0842F9D1002199601B49196000215960
1A491B48016001219960EB211966A021
196600F012F800219960164914480160
0121996001BC002800D0004712481349
086003C880F30888084703B5996A0420
0142FBD001200142F8D103BD02B51866
1866FFF7F2FF186E186E02BD00000240
000000180000070000035F0021220000
F4000018222000A00001001008ED00E0
```

for convenience the struct:
```
typedef struct __packed __aligned(4)
{
    uint32_t magic1;
    uint32_t magic2;
    uint32_t length;
    uint32_t crc32;
    uint8_t  data[];
}tFlashHeader;
``` 
I understand the struct is aligned with 4 byte chunks, but then still, the header is way bigger then the integers in the struct and I can't find any know data in the header. I would expect to find the 'magic' numbers somewhere in there. 

Would you be so kind and willing to elaborate on this?

Best regards,
Erik

**Comment by rhulme** [comment]
Hi,

As far as I can tell, it looks OK to me.  It's a little-endian processor so the numbers are stored back-to-front.  Taking the first sixteen bytes, or four 32-bit values:

```
FB5ECD8E - 0x8ECD5EFB (magic1)
A952AEC5 - 0xC5AE52A9 (magic2)
C83B0000 - 0x00003BC8 (length - 15304 bytes)
D3F26AFA - 0xFA6AF2D3 (crc32)
```
I haven't checked the the rest in detail but at least the next 16 bytes match the first 16 bytes of the usual 2nd stage bootloader.

**Reply by erikjonkers** [reply]
Ah right! check. Thanks for the clarification!
